### PR TITLE
collations/hack: cleanup dependencies

### PR DESCRIPTION
**Pull request by vmg**
## Description

Here's some cleanups for the `collations` package. I've done a small refactoring so the package is freestanding and can be used externally without depending on all of `vitess`.

While I was at it, I took a stab at the `go/hack` package and added some compatibility code so it doesn't necessarily depend on the default runtime of the Go Compiler. It should now also work on `gccgo` (if you're, like, a masochist) and `tinygo`.

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/hack/hack.go [hunk 1]
```diff
@@ -48,11 +44,3 @@ func StringBytes(s string) []byte {
...
 	return b
 }
-
-// StringClone returns a newly allocated copy of the string that doesn't share
-// its underlying memory storage.
-func StringClone(s string) string {
-	b := make([]byte, len(s))
-	copy(b, s)
-	return *(*string)(unsafe.Pointer(&b))
-}
```
Comment (vmg) [inline go/hack/hack.go:58]: This function is now part of the stdlib as of Go 1.18 :tada:

File: go/hack/hack.go [hunk 2]
```diff
@@ -32,12 +34,6 @@ func String(b []byte) (s string) {
 	return *(*string)(unsafe.Pointer(&b))
 }
 
-// StringPointer returns &s[0], which is not allowed in go
-func StringPointer(s string) unsafe.Pointer {
-	pstring := (*reflect.StringHeader)(unsafe.Pointer(&s))
-	return unsafe.Pointer(pstring.Data)
-}
-
```
Comment (vmg) [inline go/hack/hack.go:40]: This function was unused :tada:

File: go/hack/compat.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
+//go:build !gc || wasm
```
Comment (systay) [inline go/hack/compat.go:1]: shouldn't we have our header here?

File: go/hack/compat.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+package hack
+
+func String(b []byte) (s string) {
```
Comment (systay) [inline go/hack/compat.go:5]: Similarly - what about comments for these public methods?

File: go/hack/compat.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+func StringBytes(s string) []byte {
+	return []byte(s)
+}
+
+func DisableProtoBufRandomness() {}
+
+func RuntimeMemhash(b []byte, hash uint64) uint64 {
+	for i := 0; i < len(b); i++ {
+		hash *= 1099511628211
```
Comment (frouioui) [inline go/hack/compat.go:34]: Out of curiosity, what's `1099511628211`?

File: go/hack/compat.go (hunk 4 above)
Comment (vmg) [reply go/hack/compat.go:22]: Fixed!

**Review by frouioui — APPROVED** [review]

File: go/hack/compat.go (hunk 5 above)
Comment (systay) [reply go/hack/compat.go:34]: Maybe it has to do with [this](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function)?

Comment (vmg) [reply go/hack/compat.go:34]: The default prime constant in FNV1: https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/hash/fnv/fnv.go;l=36

**Pull request merged**